From abbe792e6e9dd8c78eed102cd352563d2141f9dc Mon Sep 17 00:00:00 2001
From: Peter Chen <peter.chen@freescale.com>
Date: Fri, 21 Jan 2011 17:22:44 +0800
Subject: [PATCH 647/670] ENGR00138371 usb-otg: some variables should be
 updated at otg irq

otg->host->is_b_host and otg->gadget->is_a_peripheral should be
updated at otg irq, as the host and device driver may use these two variables.

enable clk before request_irq

Signed-off-by: Peter Chen <peter.chen@freescale.com>
---
 drivers/usb/otg/fsl_otg.c | 21 ++++++++++++---------
 1 file changed, 12 insertions(+), 9 deletions(-)

diff --git a/drivers/usb/otg/fsl_otg.c b/drivers/usb/otg/fsl_otg.c
index f261461..a338f21 100644
--- a/drivers/usb/otg/fsl_otg.c
+++ b/drivers/usb/otg/fsl_otg.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (C) 2005-2010 Freescale Semiconductor, Inc. All Rights Reserved.
+ * Copyright (C) 2005-2013 Freescale Semiconductor, Inc. All Rights Reserved.
  *
  * Author: Li Yang <LeoLi@freescale.com>
  *         Jerry Huang <Chang-Ming.Huang@freescale.com>
@@ -700,11 +700,6 @@ static void fsl_otg_event(struct work_struct *work)
 	else
 		fsm->a_conn = 0;
 
-	if (otg->host)
-		otg->host->is_b_host = fsm->id;
-	if (otg->gadget)
-		otg->gadget->is_a_peripheral = !fsm->id;
-
 	if (fsm->id) {		/* switch to gadget */
 		fsl_otg_start_host(fsm, 0);
 		otg_drv_vbus(fsm, 0);
@@ -814,6 +809,14 @@ irqreturn_t fsl_otg_isr(int irq, void *dev_id)
 	fotg->fsm.id = (otg_sc & OTGSC_STS_USB_ID) ? 1 : 0;
 	otg->default_a = (fotg->fsm.id == 0);
 
+	/* is_b_host, is_a_peripheral may be used at host/gadget driver
+	 * so, assign them at irq, the otg event is scheduled too late
+	 */
+	if (otg->host)
+		otg->host->is_b_host = fotg->fsm.id;
+	if (otg->gadget)
+		otg->gadget->is_a_peripheral = !(fotg->fsm.id);
+
 	/* process OTG interrupts */
 	if (otg_int_src) {
 		if (otg_int_src & OTGSC_INTSTS_USB_ID) {
@@ -952,6 +955,9 @@ int usb_otg_start(struct platform_device *pdev)
 	p_otg->dr_mem_map = (struct usb_dr_mmap *)usb_dr_regs;
 	pdata->regs = (void *)usb_dr_regs;
 
+	if (pdata->platform_init && pdata->platform_init(pdev) != 0)
+		return -EINVAL;
+
 	gpio_id = pdata->id_gpio;
 	/* request irq */
 	if (pdata->id_gpio == 0) {
@@ -971,9 +977,6 @@ int usb_otg_start(struct platform_device *pdev)
 		return status;
 	}
 
-	if (pdata->platform_init && pdata->platform_init(pdev) != 0)
-		return -EINVAL;
-
 	clk_stopped = false; /* platform_init will open the otg clk */
 
 	/* stop the controller */
-- 
1.8.0

